From 176b29c7ff6ed96e5ec2e63a408b76747881dce4 Mon Sep 17 00:00:00 2001
From: Fam Zheng <famz@redhat.com>
Date: Mon, 4 Aug 2014 09:48:49 -0400
Subject: [virt] hyperv/storvsc: Correctly handle TEST_UNIT_READY failure

Message-id: <1407145729-25017-8-git-send-email-famz@redhat.com>
Patchwork-id: 86958
O-Subject: [RHEL-7 PATCH 7/7] drivers: scsi: storvsc: Correctly handle TEST_UNIT_READY failure
Bugzilla: 1122317
RH-Acked-by: Jason Wang <jasowang@redhat.com>
RH-Acked-by: Paolo Bonzini <pbonzini@redhat.com>

From: K. Y. Srinivasan <kys@microsoft.com>

On some Windows hosts on FC SANs, TEST_UNIT_READY can return SRB_STATUS_ERROR.
Correctly handle this. Note that there is sufficient sense information to
support scsi error handling even in this case.

Signed-off-by: K. Y. Srinivasan <kys@microsoft.com>
Reviewed-by: Hannes Reinecke <hare@suse.de>
Cc: <stable@vger.kernel.org>
Signed-off-by: Christoph Hellwig <hch@lst.de>
(cherry picked from commit 3533f8603d28b77c62d75ec899449a99bc6b77a1)
Signed-off-by: Fam Zheng <famz@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/storvsc_drv.c b/drivers/scsi/storvsc_drv.c
index d2c7cc8..c2c6515 100644
--- a/drivers/scsi/storvsc_drv.c
+++ b/drivers/scsi/storvsc_drv.c
@@ -1018,6 +1018,13 @@ static void storvsc_handle_error(struct vmscsi_request *vm_srb,
   case ATA_12:
    set_host_byte(scmnd, DID_PASSTHROUGH);
    break;
+  /*
+   * On Some Windows hosts TEST_UNIT_READY command can return
+   * SRB_STATUS_ERROR, let the upper level code deal with it
+   * based on the sense information.
+   */
+  case TEST_UNIT_READY:
+   break;
   default:
    set_host_byte(scmnd, DID_TARGET_FAILURE);
   }
-- 
1.7.1